### PR TITLE
Strong migrations for monetary fields

### PR DESCRIPTION
**Pull request by tahsinaislam**
## Link to pivotal/JIRA issue
https://codeforamerica.atlassian.net/browse/FYST-450 - related to the work on changing the monetary value field

## Is PM acceptance required? 
- No - merge after code review approval

## What was done?
- Places where integers are used in intake questions are now converted to decimals - new columns have been added 
- A rake task was created to backfill old value in demo and production once this is merged and deployed 
- future PRs will be used to change all the areas where we read/write the old columns

### CODE REVIEW

**Review by arinchoi03 — COMMENTED** [review]
more questions but looks overall good

File: app/forms/state_file/az_subtractions_form.rb [hunk 1]
```diff
@@ -10,8 +10,19 @@ class AzSubtractionsForm < QuestionsForm
...
+      else
+        attributes_to_save[:tribal_wages_amount] = tribal_wages
+      end
+
+      if armed_forces_member == "no"
+        attributes_to_save[:armed_forces_wages] = nil
+      else
+        attributes_to_save[:armed_forces_wages_amount] = armed_forces_wages
+      end
+
```
[the remark this answers is not in the document]
Comment (tahsinaislam) [reply app/forms/state_file/az_subtractions_form.rb:25]: However we want to set the tribal_wages_amount (the new column) to the tribal_wages value - yes whenever the answer is yet for the tribal_member

File: app/models/state_file1099_g.rb [hunk 2]
```diff
@@ -54,6 +58,12 @@ class StateFile1099G < ApplicationRecord
   validates :state_income_tax_withheld, numericality: { greater_than_or_equal_to: 0}
   validate :state_specific_validation
 
+  def sync_amount_columns
+    self.unemployment_compensation_amount = unemployment_compensation if unemployment_compensation.present?
+    self.federal_income_tax_withheld_amount = federal_income_tax_withheld if federal_income_tax_withheld.present?
+    self.state_income_tax_withheld_amount = state_income_tax_withheld if state_income_tax_withheld.present?
+  end
```
[the remark this answers is not in the document]
Comment (tahsinaislam) [reply app/models/state_file1099_g.rb:65]: hmmm wondering if i should do this after_save considering the validations

File: app/forms/state_file/az_subtractions_form.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (tahsinaislam) [reply app/forms/state_file/az_subtractions_form.rb:25]: ok looking at this again i realized i need to change this too haha

File: spec/models/state_file_nc_intake_spec.rb [hunk 3]
```diff
@@ -17,6 +17,7 @@
 #  df_data_imported_at               :datetime
 #  eligibility_lived_in_state        :integer          default("unfilled"), not null
 #  eligibility_out_of_state_income   :integer          default("unfilled"), not null
+#  eligibility_withdrew_529          :integer          default(0), not null
```
Comment (arinchoi03) [inline spec/models/state_file_nc_intake_spec.rb]: remove this change?

File: spec/factories/state_file_nc_intakes.rb [hunk 4]
```diff
@@ -17,6 +17,7 @@
 #  df_data_imported_at               :datetime
 #  eligibility_lived_in_state        :integer          default("unfilled"), not null
 #  eligibility_out_of_state_income   :integer          default("unfilled"), not null
+#  eligibility_withdrew_529          :integer          default(0), not null
```
Comment (arinchoi03) [inline spec/factories/state_file_nc_intakes.rb]: remove this change?

File: app/models/state_file1099_g.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (arinchoi03) [reply app/models/state_file1099_g.rb:65]: discussed together: seems like the validations run before `before_save`, so the values being set on the new columns should already be "valid" (int -> decimal type numbers so they should all be valid in the new field as well)

File: spec/factories/state_file_nc_intakes.rb (hunk 4 above)
Comment (arinchoi03) [reply spec/factories/state_file_nc_intakes.rb]: this was fixed in the subsequent merge commit!

**Review by arinchoi03 — APPROVED** [review]

File: app/controllers/state_file/questions/unemployment_controller.rb [hunk 5]
```diff
@@ -83,6 +83,9 @@ def state_file1099_params
           :federal_income_tax_withheld,
           :state_income_tax_withheld,
           :unemployment_compensation,
+          :federal_income_tax_withheld_amount,
+          :state_income_tax_withheld_amount,
+          :unemployment_compensation_amount,
```
Comment (jenny-heath) [inline app/controllers/state_file/questions/unemployment_controller.rb:88]: is this necessary given that you have the `sync_amount_columns` on the model before save?

File: app/forms/state_file/az_subtractions_form.rb [hunk 6]
```diff
@@ -10,8 +10,19 @@ class AzSubtractionsForm < QuestionsForm
...
     def save
       attributes_to_save = attributes_for(:intake)
-      attributes_to_save[:tribal_wages] = nil if tribal_member == "no"
-      attributes_to_save[:armed_forces_wages] = nil if armed_forces_member == "no"
+
+      if tribal_member == "no"
+        attributes_to_save[:tribal_wages] = nil
+      else
+        attributes_to_save[:tribal_wages_amount] = tribal_wages
+      end
```
Comment (jenny-heath) [inline app/forms/state_file/az_subtractions_form.rb:18]: i'm confused about this. what is the second case here? `tribal_member` is anything other than "no" and in that case we want to save the old column value to the new column value?

Comment (tahsinaislam) [reply app/forms/state_file/az_subtractions_form.rb:18]: yeah.. wondering if under the if tribal_member == "no" i should also set attributes_to_save[:tribal_wages_amount] = nil

**Pull request merged**